workflow/options vignettes (#458)
epiforecasts/EpiNow2: vignettes/estimate_infections_options.Rmd

For mathematical detail on the model please consult the [model definition](estimate_infections.html) vignette, and for a more general description of the use of the function, the [estimate_infections workflow](estimate_infections_workflow.html) vignette.

# Set up

We first load the _EpiNow2_ package and also the _rstan_ package that we will use later to show the differences in run times between different model options.

```r
library("EpiNow2")
#> Loading required namespace: V8
library("rstan")
#> Loading required package: StanHeaders
#> 
#> rstan version 2.26.24 (Stan version 2.26.1)
#> For execution on a local, multicore CPU with excess RAM we recommend calling
#> options(mc.cores = parallel::detectCores()).
#> To avoid recompilation of unchanged Stan programs, we recommend calling
#> rstan_options(auto_write = TRUE)
#> For within-chain threading using `reduce_sum()` or `map_rect()` Stan functions,
#> change `threads_per_chain` option:
#> rstan_options(threads_per_chain = 1)
```

In this examples we set the number of cores to use to 4 but the optimal value here will depend on the computing resources available.

```r
options(mc.cores = 4)
```

# Data

> As data set we will use an example data set that is included in the package, representing an outbreak of COVID-19 with an initial rapid increase followed by decreasing incidence.

Reviewer: ```suggestion
We will use an example data set that is included in the package, representing an outbreak of COVID-19 with an initial rapid increase followed by decreasing incidence.
```

ggplot(reported_cases, aes(x =  date, y = confirm)) +
  geom_col() +
  theme_minimal() +
  xlab("Date") +
  ylab("Cases")
```

![plot of chunk data](figure/data-1.png)

# Parameters

Before running the model we need to decide on some parameter values, in particular any delays, the generation time, and a prior on the initial reproduction number.

## Delays: incubation period and reporting delay

Delays reflect the time that passes between infection and reporting, if these exist. In this example, we assume two delays, an _incubation period_ (i.e. delay from infection to symptom onset) and a _reporting delay_ (i.e. the delay from symptom onset to being recorded as a symptomatic case). These delays are usually not the same for everyone and are instead characterised by a distribution. For the incubation period we use an example from the literature that is included in the package.

```r
incubation_period <- get_incubation_period(
  disease = "SARS-CoV-2", source = "lauer"
)
incubation_period
#> 
#>   Uncertain lognormal distribution with (untruncated) logmean 1.6 (SD 0.064) and logSD 0.42 (SD 0.069)
```

For the reporting delay, we use a lognormal distribution with mean of 2 days and 

> standard deviation of 1 day.

Reviewer: ```suggestion
standard deviation of 1 day. Note that the mean and standard deviation must be converted to the log scale.
```

```r
reporting_delay <- dist_spec(
  mean = convert_to_logmean(2, 1), mean_sd = 0,
  sd = convert_to_logsd(2, 1), sd_sd = 0, max = 10
)
reporting_delay
#> 
#>   Fixed distribution with PMF [0.11 0.48 0.27 0.093 0.029 0.0096 0.0033 0.0012 0.00045 0.00018]
```

> We can combine these delays into one by summing them up

Reviewer: ```suggestion
_EpiNow2_ provides a feature that allows us to combine these delays into one by simply summing them up
```

```

## Generation time

If we want to estimate the reproduction number we need to provide a distribution of generation times. Here again we use an example from the literature that is included with the package.

```r
generation_time <- get_generation_time(
  disease = "SARS-CoV-2", source = "ganyani"
)
generation_time
#> 
#>   Uncertain gamma distribution with (untruncated) mean 3.6 (SD 0.71) and SD 3.1 (SD 0.77)
```

## Initial reproduction number

Lastly we need to choose a prior for the initial value of the reproduction number. This is assumed by the model to be normally distributed and we can set the mean and the standard deviation. We decide to set the mean to 2 and the standard deviation to 1.

```r
rt_prior <- list(mean = 2, sd = 0.1)
```

# Running the model

We are now ready to run the model and will in the following show a number of possible options for doing so.

## Default options

By default the model uses a renewal equation for infections and a Gaussian Process prior for the reproduction number. 
Putting all the data and parameters together and tweaking the Gaussian Process to have a shorter length scale prior than the default we run.

```r

> def <- estimate_infections(reported_cases,

Reviewer: ```suggestion
def <- estimate_infections(
reported_cases,
```

# summarise results
summary(agp)
#>                                  measure                estimate
#> 1: New confirmed cases by infection date     2366 (1220 -- 4583)
#> 2:        Expected change in daily cases       Likely decreasing
#> 3:            Effective reproduction no.      0.89 (0.65 -- 1.2)
#> 4:                        Rate of growth -0.024 (-0.086 -- 0.04)
#> 5:          Doubling/halving time (days)          -29 (17 -- -8)
# elapsed time (in seconds)
get_elapsed_time(agp$fit)
#>         warmup sample
#> chain:1 27.499 31.038
#> chain:2 23.096 40.688
#> chain:3 25.129 46.259
#> chain:4 21.050 26.049
# summary plot
plot(agp)
```

![plot of chunk lower_accuracy](figure/lower_accuracy-1.png)

## Adjusting for future susceptible depletion

We might want to adjust for future susceptible depletion.
Here, we do so by setting the population to 1000000 and projecting the reproduction number from the latest estimate (rather than the default, which fixes the reproduction number to an earlier time point based on the given reporting delays).
Note that this only affects the forecasts and is done using a crude adjustment (see the [model definition](estimate_infections.html)).

```r

> dep <- estimate_infections(reported_cases,
>   generation_time = generation_time_opts(generation_time),
>   delays = delay_opts(delay),
>   rt = rt_opts(
>     prior = rt_prior,
>     pop = 1000000, future = "latest"
>   )
> )

Reviewer: ```suggestion
dep <- estimate_infections(
  reported_cases,
  generation_time = generation_time_opts(generation_time),
  delays = delay_opts(delay),
  rt = rt_opts(
    prior = rt_prior,
    pop = 1000000, future = "latest"
  )
)
```

#> variables have undefined values: gt_rev_pmf[10],The following variables have
#> undefined values: gt_rev_pmf[11],The following variables have undefined values:
#> gt_rev_pmf[12],The following variables have undefined values:
#> gt_rev_pmf[13],The following variables have undefined values:
#> gt_rev_pmf[14],The following variables have undefined values: gt_rev_pmf[15].
#> Many subsequent functions will not work correctly.
# summarise results
summary(non_parametric)
#>                                  measure                  estimate
#> 1: New confirmed cases by infection date       2333 (1759 -- 3070)
#> 2:        Expected change in daily cases                Decreasing
#> 3:            Effective reproduction no.        0.9 (0.78 -- 0.99)
#> 4:                        Rate of growth -0.023 (-0.049 -- -0.002)
#> 5:          Doubling/halving time (days)         -30 (-340 -- -14)
# elapsed time (in seconds)
get_elapsed_time(non_parametric$fit)
#>         warmup sample
#> chain:1  2.953  0.957
#> chain:2  3.467  0.933
#> chain:3  3.384  0.857
#> chain:4  3.659  0.963
# summary plot
plot(non_parametric)
```

![plot of chunk nonparametric](figure/nonparametric-1.png)

> # Running the model in production mode

Reviewer 2: This feels a little like a different vignette

Author: good idea - will pull it out
